CoreScriptsRoot/Modules/BackpackScript3D.lua

```diff
@@ -0,0 +1,259 @@
+
+local HOPPERBIN_ANGLE = math.rad(-45)
+local HOPPERBIN_ROTATION = CFrame.Angles(HOPPERBIN_ANGLE, 0, 0)
+local HOPPERBIN_OFFSET = Vector3.new(0, 0, -5)
+
+local Tools = {}
+local ToolsList = {}
+
+local BackpackScript = {}
+local topbarEnabled = false
+
+local player = game.Players.LocalPlayer
+
+local hopperbinPart = Instance.new("Part", workspace.CurrentCamera)
+hopperbinPart.Transparency = 1
+hopperbinPart.CanCollide = false
+hopperbinPart.Anchored = true
+hopperbinPart.Name = "GUI"
+local hopperbinGUI = Instance.new("SurfaceGui", player:WaitForChild("PlayerGui"))
+hopperbinGUI.Adornee = hopperbinPart
+hopperbinGUI.ToolPunchThroughDistance = 1000
+hopperbinGUI.Name = "HopperBinGUI"
+
+local verticalRange = math.rad(0)
+local horizontalRange = math.rad(0)
+
+local function DoLayout()
+	local width, height = 100, 100
+	local borderSize = (ICON_SPACING - ICON_SIZE) / 2	
+	
+	local x = borderSize
+	for i, tool in ipairs(ToolsList) do
+		local slot = Tools[tool]
+		if slot then
+			slot.icon.Position = UDim2.new(0, x, 0, 0)
+			x = x + ICON_SPACING
+		end
+	end
+	
+	width = #ToolsList * ICON_SPACING
+	height = ICON_SIZE
+	
+	hopperbinGUI.CanvasSize = Vector2.new(width, height)
+	hopperbinPart.Size = Vector3.new(width / PIXELS_PER_STUD, height / PIXELS_PER_STUD, 1)	
+	
+	verticalRange = math.atan(hopperbinPart.Size.Y / (2 * HOPPERBIN_OFFSET.Z)) + math.rad(30) * 2
+	horizontalRange = math.atan(hopperbinPart.Size.X / (2 * HOPPERBIN_OFFSET.Z)) - math.rad(20) * 2
+end
+
+local function SetTransparency(transparency)
+	for i, v in pairs(Tools) do
+		v.icon.BackgroundTransparency = transparency + 0.5
+		v.icon.ImageTransparency = transparency
+	end
+end
+
+local function AddTool(tool)
+	local slot = {}
+	Tools[tool] = slot
+	table.insert(ToolsList, tool)
+	
+	slot.tool = tool
+	slot.icon = Instance.new("ImageButton", hopperbinGUI)
+	slot.icon.Size = UDim2.new(0, ICON_SIZE, 0, ICON_SIZE)
+	slot.icon.BackgroundColor3 = Color3.new(0, 0, 0)
+	slot.icon.BorderSizePixel = 0
+	slot.icon.BorderColor3 = Color3.new(90/255, 142/255, 233/255)
+	slot.icon.Image = tool.TextureId
+	
+	slot.icon.MouseButton1Click:connect(function()
+		if not player.Character or not player.Character:FindFirstChild("Humanoid") then return end
+		
+		local in_backpack = tool.Parent == game.Players.LocalPlayer:FindFirstChild("Backpack")
+		game.Players.LocalPlayer.Character.Humanoid:UnequipTools()
+		if in_backpack then
+			game.Players.LocalPlayer.Character.Humanoid:EquipTool(tool)
+		end
+	end)
+	slot.OnEnter = function()
+		slot.icon.BorderSizePixel = 4
+	end
+	slot.OnLeave = function()
+		slot.icon.BorderSizePixel = 0
+	end
+--	slot.icon.MouseEnter:connect(slot.OnEnter)
+--	slot.icon.MouseLeave:connect(slot.OnLeave)
+	
+	DoLayout()
+end
+
+local function OnChildAdded(child)
+	if child:IsA("Tool") or child:IsA("HopperBin") then
+		if not Tools[child] then
+			AddTool(child)
+		end
+	end
+end
+
+local function RemoveTool(tool)
+	if not (Tools[tool] and Tools[tool].icon) then
```

Comment:
Even though it should always exist is there a reason we have to check the `.icon` also exists? If it does exists we can destroy it otherwise just remove the tool